src/fluree/db/connection.cljc

```diff
@@ -445,6 +446,88 @@
+                                      (util/get-first-value const/iri-address))]
+          (log/debug "Dropping commit" (-> (util/get-first commit const/iri-data)
+                                           (util/get-first-value const/iri-fluree-t)))
+          (when data-address
+            (log/debug "Deleting data" data-address)
+            (storage/delete commit-catalog data-address))
+          (when commit-address
+            (log/debug "Deleting commit" commit-address)
+            (storage/delete commit-catalog commit-address))
+          (when txn-address
+            (log/debug "Deleting txn" txn-address)
+            (storage/delete commit-catalog txn-address))
+          (recur))))))
+
+(defn drop-index-nodes
+  "Build up a list of node addresses in leaf->root order, then delete them."
+  [storage node-address]
+  (go-try
+    (loop [[address & r] [node-address]
+           addresses     (list)]
+      (if address
+        (if-let [children (->> (:children (<? (storage/read-json storage address true)))
+                               (mapv :id))]
+          (recur (into r children) (conj addresses address))
+          (recur r (conj addresses address)))
+
+        (doseq [address addresses]
+          (log/debug "Dropping node" address)
+          (storage/delete storage address))))
+    :nodes-dropped))
+
+(defn drop-index-artifacts
+  [{:keys [index-catalog] :as _conn} latest-commit]
+  (go-try
+    (let [storage       (:storage index-catalog)
+          index-address (some-> (util/get-first latest-commit const/iri-index)
+                                (util/get-first-value const/iri-address))]
+      (when index-address
+        (log/debug "Dropping index" index-address)
+        (let [{:keys [spot opst post tspo]} (<? (storage/read-json storage index-address true))
+
+              garbage-ch (garbage/clean-garbage* index-catalog index-address 0)
+              spot-ch    (drop-index-nodes storage (:id spot))
+              post-ch    (drop-index-nodes storage (:id post))
+              tspo-ch    (drop-index-nodes storage (:id tspo))
+              opst-ch    (drop-index-nodes storage (:id opst))]
+          (<? garbage-ch)
+          (<? spot-ch)
+          (<? post-ch)
+          (<? tspo-ch)
+          (<? opst-ch)
+          (<? (storage/delete storage index-address))))
+      :index-dropped)))
+
+(defn drop-ledger
+  [conn alias]
+  (go-try
+    (let [alias (if (fluree-address? alias)
+                  (nameservice/address-path alias)
+                  alias)]
+      (loop [[publisher & r] (publishers conn)]
+        (when publisher
+          (let [ledger-addr   (<? (nameservice/publishing-address publisher alias))
+                latest-commit (-> (<? (nameservice/lookup publisher ledger-addr))
+                                  json-ld/expand)]
+            (log/warn "Dropping ledger" ledger-addr)
+            (drop-index-artifacts conn latest-commit)
+            (drop-commit-artifacts conn latest-commit)
```

Comment:
This and the prior line never take `<?` and therefore despite catching exceptions in those functions they will never pull through.

That might be OK here if you decide you want everything else deleted regardless of a thrown exception, but then you could assign the result channel to a var, then before you `recur` below, do a `<?` on the var.

Also need to decide if that is OK when multiple publishers exist... should an exception in one publisher stop deletions for the others? Probably not.

Maybe need to reevaluate exception handling / reporting here. As it stands, if an exception is raised, it will stop all work for remaining publishers.

Reply:
I deliberately swallowed exceptions here since multiple nameservice records will point to different places on the commit chain, and so lower parts of the chain may already be deleted, which I think is ok. This whole operation is meant to be idempotent, so if you lose power halfway through you should be able to try again when you're back up and still delete all the artifacts.

Reply:
Then not a huge issue, but your use of `<?` implies an intention to stop processing. Maybe you just want to `<!`, or still <? but have an upper try/catch to log the exception with a `go` instead of a `go-try`.